src/plugins/data_views/server/rest_api_routes/internal/fields.ts

```diff
@@ -0,0 +1,140 @@
+ */
+
+import { createHash } from 'crypto';
+import { IRouter, RequestHandler, StartServicesAccessor, KibanaRequest } from '@kbn/core/server';
+import { unwrapEtag } from '../../../common/utils';
+import { IndexPatternsFetcher } from '../../fetcher';
+import type {
+  DataViewsServerPluginStart,
+  DataViewsServerPluginStartDependencies,
+} from '../../types';
+import type { FieldDescriptorRestResponse } from '../route_types';
+import { FIELDS_PATH as path } from '../../../common/constants';
+import { parseFields, IBody, IQuery, querySchema } from './fields_for';
+
+export function calculateHash(srcBuffer: Buffer) {
+  const hash = createHash('sha1');
+  hash.update(srcBuffer);
+  return hash.digest('hex');
+}
+
+const handler: (
+  isRollupsEnabled: () => boolean,
+  getUserId: () => (kibanaRequest: KibanaRequest) => Promise<string | undefined>
+) => RequestHandler<{}, IQuery, IBody> =
+  (isRollupsEnabled, getUserIdGetter) => async (context, request, response) => {
+    const core = await context.core;
+    const uiSettings = core.uiSettings.client;
+    const { asCurrentUser } = core.elasticsearch.client;
+    const indexPatterns = new IndexPatternsFetcher(asCurrentUser, undefined, isRollupsEnabled());
+    const {
+      pattern,
+      meta_fields: metaFields,
+      type,
+      rollup_index: rollupIndex,
+      allow_no_index: allowNoIndex,
+      include_unmapped: includeUnmapped,
+    } = request.query;
+
+    let parsedFields: string[] = [];
+    let parsedMetaFields: string[] = [];
+    try {
+      parsedMetaFields = parseFields(metaFields);
+      parsedFields = parseFields(request.query.fields ?? []);
+    } catch (error) {
+      return response.badRequest();
+    }
+
+    try {
+      const { fields, indices } = await indexPatterns.getFieldsForWildcard({
+        pattern,
+        metaFields: parsedMetaFields,
+        type,
+        rollupIndex,
+        fieldCapsOptions: {
+          allow_no_indices: allowNoIndex || false,
+          includeUnmapped,
+        },
+        ...(parsedFields.length > 0 ? { fields: parsedFields } : {}),
+      });
+
+      const body: { fields: FieldDescriptorRestResponse[]; indices: string[] } = {
+        fields,
+        indices,
+      };
+
+      const etag = calculateHash(Buffer.from(JSON.stringify(body)));
+
+      const getUserId = getUserIdGetter();
+      const userId = await getUserId(request);
+      const userHash = userId ? calculateHash(Buffer.from(userId)) : '';
+
+      const headers: Record<string, string> = {
+        'content-type': 'application/json',
+        etag,
+        vary: 'accept-encoding, user-hash',
+        'user-hash': userHash,
+      };
+
+      const cacheMaxAge = await uiSettings.get<number>('data_views:cache_max_age');
+
+      if (cacheMaxAge && fields.length) {
+        const stale = 365 * 24 * 60 * 60 - cacheMaxAge;
+        headers[
+          'cache-control'
+        ] = `private, max-age=${cacheMaxAge}, stale-while-revalidate=${stale}`;
```

Comment:
If `cacheMaxAge` is configured as `0` to disable any caching, should not it remove `stale-while-revalidate`?

Reply:
Yes, this is what this code does. The header on line 93 is used when `cacheMaxAge` is set to 0.